### PR TITLE
[ticket/15723] Rewrite gen_rand_string() and gen_rand_string_friendly()

### PR DESCRIPTION
**Pull request by rubencm**
PHPBB3-15723

Checklist:

- [x] Correct branch: master for new features; 3.2.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.2.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-15723

### CODE REVIEW

**Comment by rubencm** [comment]
unique_id returned 16 characters before, i think it should return 32 better.
In some places like salt, if you update your installation you will get a 16 characters salt, but if you make a clean installation, you will get a 32 characters salt.
https://github.com/phpbb/phpbb/blob/master/phpBB/phpbb/install/module/install_database/task/add_config_settings.php#L233
https://github.com/phpbb/phpbb/blob/master/phpBB/phpbb/db/migration/data/v310/plupload.php#L33

File: phpBB/includes/functions.php [hunk 1]
```diff
@@ -88,21 +97,25 @@ function gen_rand_string($num_chars = 8)
 */
 function gen_rand_string_friendly($num_chars = 8)
 {
-	$rand_str = bin2hex(random_bytes($num_chars + 1));
+	$range = array_merge(range('A', 'N'), range('P', 'Z'), range(1, 9));
```
Comment (Nicofuma) [inline phpBB/includes/functions.php]: you should add [a-z] too I think

File: phpBB/includes/functions.php [hunk 2]
```diff
@@ -73,8 +73,17 @@ function set_var(&$result, $var, $type, $multibyte = false)
 */
 function gen_rand_string($num_chars = 8)
 {
-	// [a, z] + [0, 9] = 36
-	return substr(strtoupper(base_convert(bin2hex(random_bytes($num_chars + 1)), 16, 36)), 0, $num_chars);
+	$range = array_merge(range('A', 'Z'), range(0, 9));
```
[the remark this answers is not in the document]
Comment (rubencm) [reply phpBB/includes/functions.php]: No, is the same characters, in the old version of the function, all characters were converted to uppercase

[the remark this answers is not in the document]
Comment (Nicofuma) [reply phpBB/includes/functions.php]: ok

**Pull request merged**